### ISSUE TITLE
Included protogen is unable to parse "oneof"

### ISSUE THREAD

**Issue by xPaw** [SteamRE/SteamKit#478]
```
tf_proto_def_messages.proto:119:9: Expected "required", "optional", or "repeated".
tf_proto_def_messages.proto:119:24: Missing field number.
```

```
steammessages_hiddevices.proto:99:9: Expected "required", "optional", or "repeated".
steammessages_hiddevices.proto:99:23: Missing field number.
```

**Comment by yaakov-h** [comment]
Nothing we can really do,

- https://github.com/mgravell/protobuf-net/issues/47
- https://github.com/mgravell/protobuf-net/issues/54

**Comment by SudoHenk** [comment]
Seems like "oneof" is not going to be implemented anytime soon. Are there any other options to generate the code from the ProtoBuffs?

**Comment by yaakov-h** [comment]
I believe `oneof` doesn't have any different wire encoding to regular fields, so in theory if you replace

```
oneof a_oneof {
    string first = 1;
    int second = 2;
}
```

with just

```
optional first = 1;
optional int second = 2;
```

then it should work, but you'd have to check in application code which of the fields is set.

**Comment by SudoHenk** [comment]
```
protobuf-net:protogen - code generator for .proto
tf_gcmessages.proto:191:16: Option "allow_alias" unknown.
tf_gcmessages.proto:7:8: Option "allow_alias" unknown.
```
Another problem, seems to be also only present in a newer version of ProtoBuf. (I'm a novice, just tried to build this project).

**Comment by mgravell** [comment]
support for `oneof` in protobuf-net has been there since June: http://blog.marcgravell.com/2017/06/protobuf-net-gets-proto3-support.html

**Comment by yaakov-h** [comment]
Thanks!

So how do we download/get/build the new protogen? Is there a convenient NuGet package?

**Comment by mgravell** [comment]
to pre-empt the next question ("how do I get protogen?") (dammit, didn't quite type fast enough - you beat me!):

- option 1: build the "protogen" tool from source - note that the args are now comparable to `protoc`
- option 2: take a look at the code in protogen's Program.cs (there isn't much) - then add a nuget reference to protobuf-net.Reflection and have fun

to pre-empt the question after that ("why isn't there a nuget package for just protogen?"):

because nuget seems to have crappy options for this kind of tool; I'm harassing MS weekly on getting improved tools for build-time codegen.

**Comment by yaakov-h** [comment]
Not even `DotNetCliTool`?

**Comment by mgravell** [comment]
honestly, I've never tried that; if that is an appropriate "fix" here, then I'm more than happy to take a look - especially if there are some reference docs somewhere

**Comment by yaakov-h** [comment]
The docs for that are [here](https://docs.microsoft.com/en-us/dotnet/core/tools/extensibility#per-project-based-extensibility), but I'm not quite sure what to make of them this late at night. 😅

**Comment by mgravell** [comment]
I'll take a look, thanks

**Comment by yaakov-h** [comment]
With an new build of protogen, the only way (it seems) to define a namespace is by using `option csharp_namespace` inside the `.proto` file. Since these protobufs are reverse-engineered, I don't think we want to mess with them.

I might have to try option 2 above and build our own generator.

**Comment by mgravell** [comment]
there are some library-specific custom options for controlling that; but
I'm not *opposed* to re-adding the options that were there previously

**Comment by xPaw** [comment]
Looks like protogen is able to do this now, just need to update it.

**Comment by yaakov-h** [comment]
I got back to this... there are bugs in protobuf-net 2.4 so we can't use that.

They seem to be fixed in 3.0 but we're now blocked waiting on this:

https://github.com/protobuf-net/protobuf-net/issues/563

**Comment by yaakov-h** [comment]
Resolved by #686.
